### PR TITLE
Allow to override chainstate options for regtest

### PR DESCRIPTION
**Pull request by stanislav-tkach**

### CODE REVIEW

File: test/functional/test_framework/test_node.py
Comment (iljakuklic) [inline test/functional/test_framework/test_node.py:110]: Cool I kind of like the command-based interface.

File: node/src/runner.rs
Comment (iljakuklic) [inline node/src/runner.rs:172]: Could the repeating pattern here be abstracted into a helper closure? Maybe it would help if `Builder` used a mutable reference instead of taking ownership and producing a new instance, otherwise some `std::mem::take` trickery may be needed.

**Review by iljakuklic — COMMENTED** [review]
Adding some initial observations, will have a proper look next week

File: node/src/runner.rs
Comment (stanislav-tkach) [reply node/src/runner.rs:172]: I'm even thinking about a macro, because both builder method and options field have the same name. I'm going to experiment with it.

File: test/functional/test_framework/test_node.py
Comment (stanislav-tkach) [reply test/functional/test_framework/test_node.py:110]: Glad to hear! I was afraid that no one would like this approach.

File: node/src/runner.rs
Comment (iljakuklic) [inline node/src/runner.rs:174]: I am not sure configuring `coin_decimals` will work correctly. It should be treated as a read-only constant or removed form chain_config for the time of being.

File: node/src/runner.rs
Comment (iljakuklic) [inline node/src/runner.rs:159]: Could these two rules just delegate to the last (most general) one, with `std::convert::identity` used as the conversion function? Less code duplication at source level but it would expand into uglier code.

**Review by iljakuklic — COMMENTED** [review]
Adding some more comments. Approving now but I think the `coin_decimals` one should be addressed or at least discussed.

**Review by iljakuklic — APPROVED** [review]

**Review by altonen — APPROVED** [review]

File: node/src/runner.rs
Comment (TheQuantumPhysicist) [reply node/src/runner.rs:174]: Even though coin_decimals is "too constant" to be treated like a variable, but I don't see an issue there. The only issue I see happening is with the reward schedule. What else could happen?

File: node/src/runner.rs
Comment (stanislav-tkach) [reply node/src/runner.rs:154]: The first branch can be delegated to the second one and it is a great idea, but I don't see how to delegate to the third branch: the types are different (`Result<T>` and `T`). Perhaps I'm missing something?

Comment (iljakuklic) [reply node/src/runner.rs:154]: Ok

File: node/src/runner.rs
Comment (stanislav-tkach) [reply node/src/runner.rs]: I have just removed this option, but I can bring it back based on the result of this discussion. 🙃

Comment (iljakuklic) [reply node/src/runner.rs]: It does not really have any effect in reward schedule, so I think it might be a rather misleading setting with poorly defined effects.

Comment (iljakuklic) [reply node/src/runner.rs]: Also it seems to me to be more of a UI concern rather than something node should be concerned with.

Comment (TheQuantumPhysicist) [reply node/src/runner.rs]: Remember that we want to test everything. Later, with rpc, we want to test how the wallet works. All this has to be tested together. It's not an option for normal ChainConfig. This is just for tests.

I'm OK with removing this option now and being able to bring it later if we have to. But I prefer to have the option than not. Having it and not using it is better than needing it and not having it.

File: node/src/options.rs
Comment (TheQuantumPhysicist) [inline node/src/options.rs:84]: Perhaps this can be moved to its own rs file. No need to mix it with the others here.

File: node/src/options.rs
Comment (TheQuantumPhysicist) [inline node/src/options.rs:52]: What about testnet? I think the model is incomplete. There can be nets other than mainnet that are not regtest.

File: node/src/options.rs
Comment (TheQuantumPhysicist) [inline node/src/options.rs:93]: One thing you can add here is the block subsidy. This is one thing we need to test in regtests. The idea is that it should be possible to submit block subsidy values in ChainConfig. There's a translation mechanism for that, where strings can be converted to a subsidy schedule. You don't have to do this in this PR, but I don't see why not. It's your call whether you want to make a separate PR for that if you find it more complex than this.

Comment (TheQuantumPhysicist) [reply node/src/options.rs:93]: Btw, there's more stuff. This is just the beginning. NetUpdates, checkpoints and more should all be passed here. But they're still not parsable as strings. That comes later.

File: node/src/runner.rs
Comment (TheQuantumPhysicist) [inline node/src/runner.rs:124]: I would make this info, not trace.

**Review by TheQuantumPhysicist — COMMENTED** [review]
The minimal requirement to get this PR ready is to support testnet configuration too. 

I also don't see tests for ChainConfig settings either, which I'm not sure about. Do we want to have tests for that?

File: node/src/runner.rs
Comment (iljakuklic) [reply node/src/runner.rs:124]: I see a newline in the log message. How does that render in the output?

Comment (stanislav-tkach) [reply node/src/runner.rs:128]: @iljakuklic Actually there are many new lines here, because the `{:#?}` formatting produces a "pretty" output by splitting it into multiple lines:
```
> RUST_LOG=info ./node -d ./eee mainnet
[2022-07-25T14:58:36Z INFO  node::runner] Starting with the following config:
     NodeConfig {
        datadir: "./eee",
        chainstate: ChainstateConfig {
            max_db_commit_attempts: 10,
            max_orphan_blocks: 512,
        },
        p2p: P2pConfig {
            bind_address: "/ip6/::1/tcp/3031",
            ban_threshold: 100,
            outbound_connection_timeout: 10,
        },
        rpc: RpcConfig {
            bind_address: 127.0.0.1:3030,
        },
    }
[2022-07-25T14:58:36Z INFO  subsystem::manager] Initializing subsystem manager mintlayer
...
```
What do you think? Should I remove the new lines from the output?

Comment (iljakuklic) [reply node/src/runner.rs:128]: Not sure. I was mainly concerned it would continue at the start of the line. However I see it's indented so should be fine. Still makes grepping a bit annoying. So at this point it's mainly aesthetic preference and I don't mind too much either way. Let's see what others think.

**Comment by TheQuantumPhysicist** [comment]
> @TheQuantumPhysicist I'm not sure about the tests. Ideally I would prefer these changes to be tested in two ways:
> 
> * Simple `assert_cmd` tests. This currently isn't trivial because `assert_cmd` don't allows to pass signals to the binary and we need that to stop the node.
> * Indirectly via functional tests. As far as I can see, the functional tests need a lot of work, so this is outside the scope of this pull request.

Actually with functional tests should happen sooner or later. If you think you can do it (in this PR or another one), I'm all for that. Again, I think this is a good opportunity to practice a little with functional tests. After all, there has to be some form of testing for these command line arguments.

**Comment by stanislav-tkach** [comment]
I suppose I can do that, but I would prefer to work on functional tests in a separate pull request.

File: node/src/runner.rs
Comment (TheQuantumPhysicist) [reply node/src/runner.rs]: Would be nice if we could convert this to a table... it's a typical thing in command line arguments to print a table on start. I can live with this for now though.

**Review by TheQuantumPhysicist — APPROVED** [review]

**Pull request merged**
